### PR TITLE
Boolean mask ops

### PR DESCRIPTION
**Pull request by rnett**
Adds custom ops for boolean masking (`tensor[mask]` and `tensor[mask] = update`).

Note that there is no Python equivalent of `BooleanMaskUpdate`, but there is in Numpy and PyTorch, so I included it.

### CODE REVIEW

**Comment by rnett** [comment]
I can't run tests locally b/c `jnitensorflow in java.library.path` errors, I'm guessing because snapshot builds were just fixed.  We'll see if the CI runs them.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/Scope.java [hunk 1]
```diff
@@ -125,6 +125,25 @@ public Scope withName(String opName) {
...
+   * <p>The default child scope name must match the regular expression {@code [A-Za-z0-9.][A-Za-z0-9_.\-]*}
+   *
+   * @param defaultName name of the sub scope if this scope's name hasn't been set.
+   * @return a new subscope
+   * @throws IllegalArgumentException if the name is invalid
+   */
+  public Scope withNameAsSubScope(String defaultName){
+    return new Scope(env, nameScope.withSubScope(nameScope.makeOpName(defaultName)), controlDependencies, deviceSpec);
+  }
+
```
Comment (karllessard) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/Scope.java:146]: It could be interesting to update our other composite ops to make use of this new method, like this one: https://github.com/tensorflow/java/blob/ef6ece686a36809a186abd7db32b0a919829b48f/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/Ones.java#L53

**Review by karllessard — APPROVED** [review]
Thanks @rnett !
